Mathlib/Analysis/Normed/Group/Ultra.lean

```diff
@@ -0,0 +1,175 @@
+@[to_additive "All triangles are isosceles in an ultrametric normed additive group."]
+lemma nnnorm_div_eq_max_of_nnnorm_div_ne_nnnorm_div (x y z : S) (h : ‖x / y‖₊ ≠ ‖y / z‖₊) :
+    ‖x / z‖₊ = max ‖x / y‖₊ ‖y / z‖₊ := by
+  simpa only [← NNReal.coe_inj, NNReal.coe_max] using
+    norm_div_eq_max_of_norm_div_ne_norm_div _ _ _ (NNReal.coe_injective.ne h)
+
+@[to_additive]
+lemma nnnorm_pow_le (x : S) (n : ℕ) :
+    ‖x ^ n‖₊ ≤ ‖x‖₊ := by
+  induction n with
+  | zero => simp
+  | succ n hn => simpa [pow_add, hn] using nnnorm_mul_le_max (x ^ n) x
+
+@[to_additive]
+lemma norm_pow_le (x : S) (n : ℕ) :
+    ‖x ^ n‖ ≤ ‖x‖ :=
+  nnnorm_pow_le x n
+
+@[to_additive]
+lemma nnnorm_zpow_le (x : S) (z : ℤ) :
+    ‖x ^ z‖₊ ≤ ‖x‖₊ := by
+  cases z <;>
+  simpa using nnnorm_pow_le _ _
+
+@[to_additive]
+lemma norm_zpow_le (x : S) (z : ℤ) :
+    ‖x ^ z‖ ≤ ‖x‖ :=
+  nnnorm_zpow_le x z
+
+end Group
+
+section CommGroup
+
+variable {M ι : Type*} [SeminormedCommGroup M] [IsUltrametricDist M]
+
+/-- Nonarchimedean norm of a product is less than or equal the norm of any term in the product.
+This version is phrased using `Finset.sup'` and `Finset.Nonempty` due to `Finset.sup`
+operating over an `OrderBot`, which `ℝ` is not.
+-/
+@[to_additive "Nonarchimedean norm of a sum is less than or equal the norm of any term in the sum.
+This version is phrased using `Finset.sup'` and `Finset.Nonempty` due to `Finset.sup`
+operating over an `OrderBot`, which `ℝ` is not. "]
+lemma _root_.Finset.Nonempty.norm_prod_le_sup'_norm {s : Finset ι} (hs : s.Nonempty) (f : ι → M) :
+    ‖∏ i ∈ s, f i‖ ≤ s.sup' hs (‖f ·‖) := by
+  simp only [Finset.le_sup'_iff]
+  refine Finset.Nonempty.cons_induction ?_ ?_ hs
+  · simp
+  rintro a s ha _ ⟨b, hb, IH⟩
+  simp only [Finset.prod_cons, Finset.mem_cons, exists_eq_or_imp]
+  refine (le_total ‖∏ i ∈ s, f i‖ ‖f a‖).imp ?_ ?_ <;> intro h
+  · exact (norm_mul_le_max _ _).trans (by simp [h])
+  · exact ⟨b, hb, (norm_mul_le_max _ _).trans (by simpa [h] using IH)⟩
+
+/-- Nonarchimedean norm of a product is less than or equal the norm of any term in the product. -/
+@[to_additive "Nonarchimedean norm of a sum is less than or equal the norm of any term in the sum."]
+lemma _root_.Finset.nnnorm_prod_le_sup_nnnorm (s : Finset ι) (f : ι → M) :
+    ‖∏ i ∈ s, f i‖₊ ≤ s.sup (‖f ·‖₊) := by
+  rcases s.eq_empty_or_nonempty with rfl|hs
+  · simp
+  rw [← Finset.sup'_eq_sup hs, ← NNReal.coe_le_coe,
+    Finset.comp_sup'_eq_sup'_comp hs toReal coe_mono.map_sup]
+  simpa only [coe_nnnorm', Function.comp_apply, Finset.le_sup'_iff] using
+    hs.norm_prod_le_sup'_norm f
+
+/-- Nonarchimedean norm of a product is less than or equal the norm of any term in the product. -/
+@[to_additive "Nonarchimedean norm of a sum is less than or equal the norm of any term in the sum."]
+lemma _root_.Fintype.nnnorm_prod_le_sup_univ_norm [Fintype ι] (f : ι → M) :
+    ‖∏ i, f i‖₊ ≤ Finset.univ.sup (‖f ·‖₊) := by
+  simpa using Finset.univ.nnnorm_prod_le_sup_nnnorm f
```

Comment:
Are this lemma and the next needed? By definition, `∏ x, f x` is just syntactic sugar for `Finset.prod Finset.univ f`, so these lemmas are special cases of the ones above. (If you squeeze the simpa, it's not actually using any lemmas at all.)

Reply:
We have a variety of lemmas that match the Fintype prod/sum syntactically, like `Fintype.prod_sum`. I make these basic lemmas to make life easier for the downstream consumer.

Reply:
I feel that's a slightly different situation. For `Fintype.prod_sum`, the more general lemmas `Finset.prod_univ_sum` and `Finset.prod_sum` are both slightly more awkward to state than the fintype version, so if you used the finset lemmas applied to `Finset.univ`, you would be left with a `Finset.attach` or `Fintype.piFinset` to be tidied away afterwards. In the case at hand, the statement of the Fintype lemma is literally identical to the Finset lemma applied to `univ` (even at the very weak transparency level used by `rw`); so I don't think there's any actual gain for the downstream consumer.